### ISSUE TITLE
Limit users sign in

### ISSUE THREAD

**Issue by rizlas** [Timshel/vaultwarden#10]
Hi,
first of all thank you for implementing SSO support to vaultwarden. I would like to ask: how can we limit users to sign in to vaultwarden using SSO? I'm currently using keycloak provider and even with authorization enabled and declared policies, users can still login.

Is this unsupported?

Is it also possible or will be possible to set the claim used for users username?

Tested with docker image build from sso-support branch (latest commit 94402f782175c9b9a65fb361a31b73715cc0ae03).

Thank you

---

Just for cross reference: https://github.com/dani-garcia/vaultwarden/pull/3899

**Comment by Timshel** [comment]
> I would like to ask: how can we limit users to sign in to vaultwarden using SSO? I'm currently using keycloak provider and even with authorization enabled and declared policies, users can still login.

I'm unsure how you are testing, but switching to `SSO_ONLY=true` should prevent login using Master password.
https://github.com/Timshel/vaultwarden/blob/94402f782175c9b9a65fb361a31b73715cc0ae03/.env.template#L363

> Is it also possible or will be possible to set the claim used for users username?

No, for now it uses the `preferred_username` and it's not planned to allow to change the claim.

**Comment by rizlas** [comment]
> I'm unsure how you are testing, but switching to `SSO_ONLY=true` should prevent login using Master password.

It is indeed preventing login with master password. 
Let's say I want only a subset of my user to be able to login via SSO. In keycloak I implemented this https://stackoverflow.com/a/54374808. Trying to evaluate the rules set, keycloak says that my user cannot log in, but on vaultwarden I can log in.

**Comment by Timshel** [comment]
There is no mechanism in VaultWarden to whitelist only a subset of users; if keycloak return a valid authorization code then the user will be logged-in.

I had a quick look at the Stackoverflow you linked it does appear to be a solution purely implemented in Keycloak.
I have only basic knowledge of Keycloak so I won't be able to help you with it, good luck.

**Comment by rizlas** [comment]
Keycloak side is pretty straightforward as the link says.

Would be great to have the possibility to filter out users based on attributes coming from KC. As this [comment](https://stackoverflow.com/questions/54305880/how-can-i-restrict-client-access-to-only-one-group-of-users-in-keycloak/54374808#comment111720218_54374808) says,  policy enforcement should be handled client side.

**Comment by Timshel** [comment]
The point of the SSO is to externalize the handling of the user authentication so to me the filtering you want should be done in Keycloak and not by the client, the first answer https://stackoverflow.com/a/64636667 appear to do this.

In the case of an SSO you do not control like for example using gitlab I can understand the use and logic of filtering in the client. But this feature is not supported and not planned at the moment.

**Comment by Kofl** [comment]
Can agree here, the filtering should be done on the Keycloak or in general at the identity provider side.
Worked fine with Entra ID in my case, where you selective assign users to the SSO app to allow login.

**Comment by rizlas** [comment]
In keycloak is not possible, you risk to mess up things. 

https://github.com/sventorben/keycloak-restrict-client-auth

Didn't tested this extension, could be a way. But if you listen to the video, both guys said that authz should be handled by the application.

**Comment by Timshel** [comment]
To moderate a bit my response:

There is a logic of and Admin vs default user in VaultWarden but at the moment it's not handled by the sso.
If support for such role is added then having the logic that  a user without a valid role cannot access the app would probably be something I would add (with probably a config to otherwise fallback to user).

I still believe that if only handled in the app it's kind of a bad UX since you'll redirect the user only to reject him. You could argue that it's Authentication vs Authorization but Keycloak would still provide both.

Anyway, I will probably have a look at it since I'm testing adding groups/organization mapping in [Timshel/vaultwarden/main](https://github.com/Timshel/vaultwarden).
But don´t expect it to land in the PR since it would increase its scope which is already a hurdle for it's review/integration.

**Comment by rizlas** [comment]
> Then it's a keycloak limitation in relation to other IdPs?

Yes, I tested an answer from SO and I was able to block login. How? KC wasn't sending email scope. I didn't tested it thoroughly, but probably I removed the use of email scope to all my clients.


> Anyway, I will probably have a look at it since I'm testing adding groups/organization mapping in [Timshel/vaultwarden/main](https://github.com/Timshel/vaultwarden). But don´t expect it to land in the PR since it would increase its scope which is already a hurdle for it's review/integration.

I can't ask more. Having SSO is already a big thing and achievement for vaultwarden project. Of course mapping groups and organization will be another big step, I agree that is out of the scope of this PR.

I opened an issue but the purpose was actually to ask for clarification and at this point a feature request.

Thank you again 🔝

**Comment by Timshel** [comment]
@rizlas hey so added `admin` support using `role` to the `main` branch : [Timshel/vaultwarden#role-mapping](https://github.com/Timshel/vaultwarden#role-mapping).

There is a parameter `SSO_ROLES_DEFAULT_TO_USER` which will block login if no `user` or `admin` role is present but the flow is not really nice, the user will be stuck on a loading page with no error message.  

I'm a bit unsure why the error is not displayed but since it's bitwarden client code might not be able to do better :(

**Comment by rizlas** [comment]
Thank you Tim! I'll test it out

**Comment by Timshel** [comment]
Hey,
Made a [patch](https://github.com/Timshel/oidc_web_builds/blob/master/oidc_sso_errors.patch) to the frontend to correctly display errors of the `/connect/token` endpoint.

So with latest version of `main` and the `oidc_web_vault` running the `override` version if the user does not have the correct role it will redirect him to the start of the flow and display an error message.

**Comment by rizlas** [comment]
Hi, currently testing. It is working but I'm failing to test the error message. I built this image https://github.com/Timshel/vaultwarden/blob/main/docker/Dockerfile.debian. 

Is it correct?

**Comment by rizlas** [comment]
## Implementation

Implementation based on role is working fine. Tested adding a user with role "user" or "admin, "user" and admin".
This is my KC setup for reference:

Two groups:

![immagine](https://github.com/Timshel/vaultwarden/assets/8522635/69c8d68c-03cc-4bbc-ae31-a877fb3d7ef7)
![immagine](https://github.com/Timshel/vaultwarden/assets/8522635/98068a7b-a30f-4ca7-9daa-b6c858e874a6)

Which hook into the client's role mapping:

![immagine](https://github.com/Timshel/vaultwarden/assets/8522635/4fac47a4-4180-4d06-8742-660245a30b58)

Client scope role as default:

![immagine](https://github.com/Timshel/vaultwarden/assets/8522635/dd138dfc-1c97-4ca2-af62-dfbe7d08720e)

---

## Group Organization invitation

https://github.com/Timshel/vaultwarden/tree/main#grouporganization-invitation-mapping

Didn't catch how this env var works. Can you elaborate more on this? Also, i think that `SSO_ORGANIZATIONS_INVIT` is mistyped.

---

## Notice

During testing I noticed that my SSO session was expired (checked keycloak side) but I was still able to login in my vault using master password.

**Comment by Timshel** [comment]
> Hi, currently testing. It is working but I'm failing to test the error message. I built this image https://github.com/Timshel/vaultwarden/blob/main/docker/Dockerfile.debian.

This is the correct image but are you running the modified front-ent ? (`-e SSO_FRONTEND='override'`), otherwise the error is not displayed on the client and stay blocked on a loading.

>Didn't catch how this env var works. Can you elaborate more on this?

I improved a bit the doc: https://github.com/Timshel/vaultwarden/tree/main#grouporganization-invitation-mapping

> During testing I noticed that my SSO session was expired (checked keycloak side) but I was still able to login in my vault using master password.

Hum maybe an expiry information which is not passed will check thx.

**Comment by rizlas** [comment]
> This is the correct image but are you running the modified front-ent ? (`-e SSO_FRONTEND='override'`), otherwise the error is not displayed on the client and stay blocked on a loading.

I wasn't but I'm experiencing the same result. After that another error, I was pushing the value inside .env file 😅 
Now is working. I'd like to point out that, the popup message will need: internationalization and more display time. It just vanish in a few seconds.

> I improved a bit the doc: https://github.com/Timshel/vaultwarden/tree/main#grouporganization-invitation-mapping

Thank you, I'll have a look at it.

**Comment by Timshel** [comment]
> internationalization and more display time. It just vanish in a few seconds. 

For the Internationalization  part some of it could be done but only for the web front-end not for the Android and IOS app.
So I probably won't work on it.

For the display time it could be extended a bit, will check if there is an easy place to change it.

**Comment by Timshel** [comment]
Hey, closing because I believe it's working ?
If anything else don't hesitate to open another issue :).

**Comment by rizlas** [comment]
Yeah it is working. I would like to test one more thing about the organization invitation and user management. I'll report here one I tested it.
Thank you

**Comment by rizlas** [comment]
Hi,
I did my final tests. Here's a summary:

- https://github.com/Timshel/vaultwarden/blob/main/README.md?plain=1#L34 -> Check if an Organization with a matching name (case sensitive) exist and the user is not part of it. I know that case sensitivity is already stated above, but better to be clear imho.
- Logging out from admin panel doesn't end the session.

    1. Click logout from admin panel
    2. Red banner is shown
    3. Go to base url (e.g. vaultwarden.myfqdn.tld)
    4. Master password form appears instead of "Login with SSO" form.
    5. Session still open KeyCloak side
     
    Also it would be nice to logout from admin panel and automatically redirect to "Login with SSO" form.
    
- Remove user from a group doesn't remove it from organization. Is this intended?
- https://github.com/Timshel/vaultwarden/blob/main/README.md?plain=1#L52 Add default value (I assume email profile)
- https://github.com/Timshel/vaultwarden/blob/main/README.md?plain=1#L54 Add default value (I assume it is /groups)

**Comment by Timshel** [comment]
> Logging out from admin panel doesn't end the session.

Yes Admin session is separate from Vault session. More integration would be possible but would need to be discussed with the VaultWarden maintainers.

> Session still open KeyCloak side

For admin cf previous. But I'll not it's the same when you logout from the Vault but since the front does not call the server can't do much :( (some documentation mention they are working on SLO but nothing atm I believe).

> Remove user from a group doesn't remove it from organization. Is this intended?

It's just not implemented, should not be an issue, there is a `UserOrgStatus.Revoked` status which would probably be a match but would need to took at it. Might have a look but no guaranty ^^

> In your opinion is it safe?

Probably different level of consideration.

On pure security: I believe it is but at the same time I'm not working in security ^^. 
 - Adding the SSO add an opportunity for a MITM attack but `SSO_KEY_FILEPATH` is available is you want to reinforce against that. 
 - It adds additional exchange with potential for replayability but a `nonce` is used to protect against that.
 - Review should help in case I made some mistake ^^.
 - In case of issues it might give easier access to the key and private_key of the user but the Master password would still be needed.

As for the aspect of maintainability it will depends of how much risk you can tolerate and how acceptable the different degraded solution are acceptable to you.

On the base SSO feature which is included in the [PR](https://github.com/dani-garcia/vaultwarden/pull/3899) I would expect they will probably be merged but at the same time it's in what I would call a merge-able state since September (recent issue with the session lifetime could have be fixed post merge for example). Blackdex one of maintainer is reviewing the PR lately but I have no insight if it means it might be merged in the coming month or later.

Once merged then there is the issue of the additional role and groups features, I would then open another PR but I have no idea if there will be an interest in the features.

Then there is the issue of the front-end, most of the change in the `override` version will probably NOT be integrated, I'm hoping the fix for the organization invitation will since it became simpler (initially it was not expected to be).

In general keeping the code up to date has not been an issue but in the end it means depending on two set of maintainers one composed of only me ^^. So I would balance it against how much of an issue is-it if I disappear and you are forced to either maintain it or revert to a default VaultWarden (which is trivial).

**Comment by rizlas** [comment]
Thank you, great explanation of everything!
I merely think that, at this point and after all your work, a shared roadmap would be nice. I'm sure that SSO PR is going to be merged sooner or later, people is asking and waiting for it, only a fool could ever close without merging a PR like that.

I also rate this issue and the development associated, a nice add on that will complete SSO introduction. Maybe it is worth to start discussing it with BlackDex?

**Comment by Timshel** [comment]
> I also rate this issue and the development associated, a nice add on that will complete SSO introduction. Maybe it is worth to start discussing it with BlackDex?

I have trouble getting answers for the current open PR so anything beyond that is not worth the effort at the moment.

**Comment by rizlas** [comment]
Yeah, of course not now and not in the current PR.
